src/Utils.php

```diff
@@ -700,30 +686,69 @@ function _wf_civicrm_api3_contribution_transact_spec(&$params) {
    *   contribution of created or updated record (or a civicrm error)
    */
   function wf_civicrm_api3_contribution_transact($params) {
-    // CRM_Core_Error::deprecatedFunctionWarning('The contibution.transact api is unsupported & known to have issues. Please see the section at the bottom of https://docs.civicrm.org/dev/en/latest/financial/OrderAPI/ for getting off it');
-    // Set some params specific to payment processing
-    // @todo - fix this function - none of the results checked by civicrm_error would ever be an array with
-    // 'is_error' set
-    // also trxn_id is not saved.
-    // but since there is no test it's not desirable to jump in & make the obvious changes.
-    $params['payment_processor_mode'] = empty($params['is_test']) ? 'live' : 'test';
-    $params['amount'] = $params['total_amount'];
-    if (!isset($params['net_amount'])) {
-      $params['net_amount'] = $params['amount'];
-    }
-    if (!isset($params['invoiceID']) && isset($params['invoice_id'])) {
+    // Start with the same parameters as Contribution.transact.
+    $params['contribution_status_id'] = 'Pending';
+    if (!isset($params['invoice_id']) && !isset($params['invoiceID'])) {
+      // Set an invoice_id here if you have not already done so.
+      // Potentially Order api should do this https://lab.civicrm.org/dev/financial/issues/78
+      $params['invoice_id'] = md5(uniqid(rand(), TRUE));
+    }
+    if (isset($params['invoice_id']) && !isset($params['invoiceID'])) {
+      // This would be required prior to https://lab.civicrm.org/dev/financial/issues/77
       $params['invoiceID'] = $params['invoice_id'];
     }
+    elseif (!isset($params['invoice_id']) && isset($params['invoiceID'])) {
+      $params['invoice_id'] = $params['invoiceID'];
+    }
+
+    $order = civicrm_api3('Order', 'create', $params);
+    try {
+      // Use the Payment Processor to attempt to take the actual payment. You may
+      // pass in other params here, too.
+      $propertyBag = new \Civi\Payment\PropertyBag();
+      $propertyBag->mergeLegacyInputParams($params);
+      $propertyBag->setContributionID($order['id']);
+      // @fixme: Class \Civi\Payment\CRM_Utils_Money not found (fixed in 5.28 via https://github.com/civicrm/civicrm-core/pull/17505
+      //   But note we still return a localized format
+      // $propertyBag->setAmount($params['total_amount']);
+
+      $reflectionClass = new ReflectionClass($propertyBag);
+      $reflectionProperty = $reflectionClass->getProperty('props');
+      $reflectionProperty->setAccessible(TRUE);
```

Comment:
What is `reflectionClass`?

Reply:
Have now added a comment to the code. It's a way of accessing protected vars/methods in a class because https://github.com/civicrm/civicrm-core/pull/17507 has not (yet) been merged... We need PropertyBag as an array and this gets it.